### PR TITLE
Migrate SqrtPriceMath tests from HardHat to Foundry

### PR DESCRIPTION
**Pull request by gitcoindev**
## Related Issue
This pull request migrates SqrtPriceMath tests from HardHat to Foundry (41 test cases),  a part of #315 .

## Description of changes
* migrated 41 SqrtPriceMath tests to Foundry
* refactored constants to separate library to be re-used by multiple tests
* removed HardHat implementation after migratiopn

```
Running 41 tests for test/foundry-tests/SqrtPriceMath.t.sol:SqrtPriceMathTestTest
[PASS] test_getAmount0Delta_gasCostForAmount0WhereRoundUpIsFalse() (gas: 6795)
[PASS] test_getAmount0Delta_gasCostForAmount0WhereRoundUpIsTrue() (gas: 6959)
[PASS] test_getAmount0Delta_returns0IfLiquidityIs0() (gas: 6634)
[PASS] test_getAmount0Delta_returns0IfPricesAreEqual() (gas: 6634)
[PASS] test_getAmount0Delta_returns0_1Amount1ForPriceOf1To1_21() (gas: 8814)
[PASS] test_getAmount0Delta_worksForPricesThatOverflow() (gas: 9380)
[PASS] test_getAmount1Delta_gasCostForAmount1WhereRoundUpIsFalse() (gas: 6805)
[PASS] test_getAmount1Delta_gasCostForAmount1WhereRoundUpIsTrue() (gas: 6902)
[PASS] test_getAmount1Delta_returns0IfLiquidityIs0() (gas: 6635)
[PASS] test_getAmount1Delta_returns0IfPricesAreEqual() (gas: 6623)
[PASS] test_getAmount1Delta_returns0_1Amount1ForPriceOf1To1_21() (gas: 8855)
[PASS] test_getNextSqrtPriceFromInput_amountInGreaterThanType_uint96_maxAndZeroForOneEqualsTrue() (gas: 6996)
[PASS] test_getNextSqrtPriceFromInput_anyInputAmountCannotUnderflowThePrice() (gas: 6852)
[PASS] test_getNextSqrtPriceFromInput_canReturn1WithEnoughAmountInAndZeroForOneEqualsTrue() (gas: 6777)
[PASS] test_getNextSqrtPriceFromInput_inputAmountOf0_1Currency0() (gas: 7222)
[PASS] test_getNextSqrtPriceFromInput_inputAmountOf0_1Currency1() (gas: 7036)
[PASS] test_getNextSqrtPriceFromInput_returnsInputPriceIfAmountInIsZeroAndZeroForOneEqualsFalse() (gas: 6658)
[PASS] test_getNextSqrtPriceFromInput_returnsInputPriceIfAmountInIsZeroAndZeroForOneEqualsTrue() (gas: 6400)
[PASS] test_getNextSqrtPriceFromInput_returnsTheMinimumPriceForMaxInputs() (gas: 6925)
[PASS] test_getNextSqrtPriceFromInput_revertsIfInputAmountOverflowsThePrice() (gas: 9303)
[PASS] test_getNextSqrtPriceFromInput_revertsIfLiquidityIsZero() (gas: 9116)
[PASS] test_getNextSqrtPriceFromInput_revertsIfPriceIsZero() (gas: 9129)
[PASS] test_getNextSqrtPriceFromInput_zeroForOneEqualsFalseGas() (gas: 6980)
[PASS] test_getNextSqrtPriceFromInput_zeroForOneEqualsTrueGas() (gas: 7155)
[PASS] test_getNextSqrtPriceFromOutput_outputAmountOf0_1Currency0() (gas: 6944)
[PASS] test_getNextSqrtPriceFromOutput_outputAmountOf0_1Currency1() (gas: 7298)
[PASS] test_getNextSqrtPriceFromOutput_puzzlingEchidnaTest() (gas: 9101)
[PASS] test_getNextSqrtPriceFromOutput_returnsInputPriceIfAmountInIsZeroAndZeroForOneEqualsFalse() (gas: 6627)
[PASS] test_getNextSqrtPriceFromOutput_returnsInputPriceIfAmountInIsZeroAndZeroForOneEqualsTrue() (gas: 6869)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfAmountOutIsImpossibleInOneForZeroDirection() (gas: 9120)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfAmountOutIsImpossibleInZeroForOneDirection() (gas: 9165)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfLiquidityIsZero() (gas: 9137)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfOutputAmountIsExactlyTheVirtualReservesOfCurrency0() (gas: 9101)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfOutputAmountIsExactlyTheVirtualReservesOfCurrency1() (gas: 9165)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfOutputAmountIsGreaterThanTheVirtualReservesOfCurrency0() (gas: 9152)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfOutputAmountIsGreaterThanTheVirtualReservesOfCurrency1() (gas: 9165)
[PASS] test_getNextSqrtPriceFromOutput_revertsIfPriceIsZero() (gas: 9085)
[PASS] test_getNextSqrtPriceFromOutput_succeedsIfOutputAmountIsJustLessThanTheVirtualReservesOfCurrency1() (gas: 6579)
[PASS] test_getNextSqrtPriceFromOutput_zeroForOneEqualsFalseGas() (gas: 7285)
[PASS] test_getNextSqrtPriceFromOutput_zeroForOneEqualsTrueGas() (gas: 6968)
[PASS] test_swapComputation_sqrtPTimessqrtQOverflows() (gas: 9374)
Test result: ok. 41 passed; 0 failed; finished in 4.84ms
```

### CODE REVIEW

**Comment by gitcoindev** [comment]
I force pushed the branch after signing commits, no changes were made. PR ready for the review.

**Comment by gitcoindev** [comment]
Rebased to latest main branch.

File: test/foundry-tests/SqrtPriceMath.t.sol
Comment (ewilz) [inline test/foundry-tests/SqrtPriceMath.t.sol:1]: this is how we do gas snapshotting in forge, will need  the same gas snapshots https://github.com/Uniswap/v4-core/blob/ff1dc91a496a329ada6c410a7b1b3cd8cf782160/test/foundry-tests/PoolManager.t.sol#L339-L341

File: test/foundry-tests/SqrtPriceMath.t.sol [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+    }
+
+    function expandTo18Decimals(uint256 a) internal pure returns (uint256) {
+        uint256 multiplier = 10 ** 18;
+        return a * multiplier;
+    }
+
+    function test_getNextSqrtPriceFromInput_revertsIfPriceIsZero() public {
+        vm.expectRevert();
+        sqrtPriceMath.getNextSqrtPriceFromInput(0, 0, expandTo18Decimals(1) / 10, false);
```
Comment (ewilz) [inline test/foundry-tests/SqrtPriceMath.t.sol:24]: can get rid of `expandTo18Decimals()` and just write `1 ether`

File: test/foundry-tests/SqrtPriceMath.t.sol [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+    function test_getNextSqrtPriceFromInput_returnsInputPriceIfAmountInIsZeroAndZeroForOneEqualsFalse() public {
+        uint160 price = Constants.SQRT_RATIO_1_1;
+        uint128 liquidity = 1;
+
+        assertEq(sqrtPriceMath.getNextSqrtPriceFromInput(price, liquidity, 0, false), price);
+    }
+
+    function test_getNextSqrtPriceFromInput_returnsTheMinimumPriceForMaxInputs() public {
+        uint160 sqrtP = 2 ** 160 - 1;
```
Comment (ewilz) [inline test/foundry-tests/SqrtPriceMath.t.sol:64]: can jut use `type(uint160).max` for max inputs in solidity

File: test/foundry-tests/SqrtPriceMath.t.sol [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        uint128 liquidity = Constants.MAX_UINT128;
+        uint256 maxAmountNoOverflow = Constants.MAX_UINT256 - Constants.MAX_UINT128 << 96 / sqrtP;
+
+        assertEq(sqrtPriceMath.getNextSqrtPriceFromInput(sqrtP, liquidity, maxAmountNoOverflow, true), 1);
+    }
+
+    function test_getNextSqrtPriceFromInput_inputAmountOf0_1Currency1() public {
+        uint160 sqrtP = Constants.SQRT_RATIO_1_1; // sqrtP = encodeSqrtPriceX96(1, 1)
```
Comment (ewilz) [inline test/foundry-tests/SqrtPriceMath.t.sol:72]: don't need javascript reference comment here

Comment (gitcoindev) [reply test/foundry-tests/SqrtPriceMath.t.sol:72]: @ewilz thank you for the review! I will incorporate changes and push the updates soon.

**Comment by gitcoindev** [comment]
hi @ewilz I applied the suggested changes, the pr is ready for another review round.

File: test/foundry-tests/SqrtPriceMath.t.sol [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+    function test_getNextSqrtPriceFromInput_zeroForOneEqualsFalseGas() public {
+        uint160 sqrtP = Constants.SQRT_RATIO_1_1;
+
+        snapStart("getNextSqrtPriceFromInput_zeroForOneEqualsTrueGas");
+        sqrtPriceMath.getGasCostOfGetNextSqrtPriceFromInput(sqrtP, uint128(1 ether), 0.1 ether, false);
+        snapEnd();
+    }
+
+    // #getNextSqrtPriceFromOutput
```
Comment (ewilz) [inline test/foundry-tests/SqrtPriceMath.t.sol:117]: not needed

File: test/foundry-tests/SqrtPriceMath.t.sol [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+    function test_getNextSqrtPriceFromOutput_zeroForOneEqualsFalseGas() public {
+        uint160 sqrtP = Constants.SQRT_RATIO_1_1;
+
+        snapStart("getNextSqrtPriceFromOutput_zeroForOneEqualsFalseGas");
+        sqrtPriceMath.getGasCostOfGetNextSqrtPriceFromOutput(sqrtP, uint128(1 ether), 0.1 ether, false);
+        snapEnd();
+    }
+
+    // #getAmount0Delta
```
Comment (ewilz) [inline test/foundry-tests/SqrtPriceMath.t.sol:248]: not needed

**Comment by ewilz** [comment]
cool! can you please delete `test/_snapshots_` run `yarn run snapshots` to regenerate that folder to account for the snaps you deleted, and also run `forge test` to generate the forge snapshots?

File: test/foundry-tests/SqrtPriceMath.t.sol (hunk 4 above)
Comment (gitcoindev) [reply test/foundry-tests/SqrtPriceMath.t.sol]: deleted

File: test/foundry-tests/SqrtPriceMath.t.sol (hunk 5 above)
Comment (gitcoindev) [reply test/foundry-tests/SqrtPriceMath.t.sol]: deleted

**Comment by gitcoindev** [comment]
> cool! can you please delete `test/_snapshots_` run `yarn run snapshots` to regenerate that folder to account for the snaps you deleted, and also run `forge test` to generate the forge snapshots?

done added at https://github.com/Uniswap/v4-core/pull/317/commits/a72796e724d664dbb2c7b4f60c6fc6d078e1062c

**Comment by gitcoindev** [comment]
aha! I noticed one copy-paste error, as one snapshot was duplicated, fixing asap

**Comment by gitcoindev** [comment]
@ewilz new forge gas snapshots were generated and pushed, PR is ready for the new review round.

**Comment by gitcoindev** [comment]
@ewilz gas costs look much better now, see https://github.com/Uniswap/v4-core/pull/317/commits/9ca00b0bf4736935d89dc880195b2105d9c48db6

**Comment by gitcoindev** [comment]
Thank you for merging! I will apply the same changes to https://github.com/Uniswap/v4-core/pull/320 and https://github.com/Uniswap/v4-core/pull/322

**Pull request merged**